[dnf5] Fix --help in subcommands, refactor microdnf's Command objects

dmach opened the pull request.

Commit:
microdnf: Remove unused configure() and pre_configure()

Commit:
argparse: Print all available commands, options and arguments by default

Commit:
argparse: Better align printed help

Commit:
cli: Create Session and Command classes

Commit:
microdnf: Inherit Context from Session, use the new Command class

Commit:
microdnf: Improve command short descriptions

Commit:
microdnf: Remove redundant opt/arg header descriptions

Commit:
microdnf: Add distro-sync command stub

Commit:
microdnf: Add --unneeded option stub to the remove command

Commit:
microdnf: Add swap command stub

Commit:
microdnf: Add search command stub

jrohel commented on libdnf-cli/session.cpp:
```diff
@@ -0,0 +1,48 @@
+
+Command::Command(Session & session) : session{&session} {
+    // create a new root command; owned by the arg_parser
+    argument_parser_command = session.get_argument_parser().add_new_command("root");
```
The idea was that the root command name is the name of the application, eg "microdnf". What to extend the Command constructor with the `name` argument? And the application sets the `name`.

dmach replied:
Done. I've used `program_name` to make it more explicit.

jrohel commented on microdnf/commands/command.hpp:
```diff
@@ -27,8 +27,6 @@ class Context;
 public:
     virtual void set_argument_parser(Context &) {}
-    virtual void pre_configure(Context &) {}
-    virtual void configure(Context &) {}
```
The idea of these methods (`pre_configure` and `configure`) was to allow the selected command to do some things before and after loading the configuration. And in the future, plug-in commands could use that. But it has not been used yet. Well. We can remove these methods and add them back as needed in the future.

jrohel commented on libdnf-cli/session.cpp:
```diff
@@ -0,0 +1,48 @@
+    session.get_argument_parser().set_root_command(argument_parser_command);
+
+    // set the created root command as the selected
+    // the value will get overwriiten with a subcommand during parsing the arguments
```
typo: overwritten

dmach replied:
Fixed.

jrohel commented on include/libdnf-cli/session.hpp:
```diff
@@ -0,0 +1,115 @@
+    explicit Command(Session & session);
+    virtual ~Command() = default;
+
+    /// Run the command. Must be overriden in inherited classes.
```
Really "must" ? A default empty implementation is provided.
Another solution is to use pure virtual method. It "must" then be implemented in a derived class:
`virtual void run() = 0;`

jrohel commented on include/libdnf-cli/session.hpp:
```diff
@@ -0,0 +1,115 @@
+    /// @return Pointer to the Session.
+    ///         The returned pointer must **not** be freed manually.
+    /// @since 5.0
+    Session * get_session() { return session; }
```
`noexcept`. If the `session` is not set, it returns `nullptr`.
Or return a reference and throw an exception when the `session` is not set.

dmach replied:
I'll use `noexcept` and return a reference.

jrohel commented on include/libdnf-cli/session.hpp:
```diff
@@ -0,0 +1,115 @@
+    /// @return Pointer to the parent Command. Root command returns null because it has no parent.
+    ///         The returned pointer must **not** be freed manually.
+    /// @since 5.0
+    Command * get_parent_command() { return parent_command; }
```
`noexcept`. If the `patrent_command` is not set, it returns `nullptr`.

jrohel commented on include/libdnf-cli/session.hpp:
```diff
@@ -0,0 +1,115 @@
+    /// @return Pointer to the underlying argument parser command.
+    ///         The returned pointer must **not** be freed manually.
+    /// @since 5.0
+    libdnf::cli::ArgumentParser::Command * get_argument_parser_command() { return argument_parser_command; }
```
`noexcept`. If the `argument_parser_comman` is not set, it returns `nullptr`.

jrohel reviewed (changes requested):
Please, split the  first commit "cli: Print help of selected command" to two commits:
- argparse: Add `ArgumentParser::get_selected_command` method  // (without the change in argument_parser.cpp at line 382/384)
- use `ArgumentParser::get_selected_command` in microdnf and dnfdaemon-client

Commit:
argparse: Store the selected command in the argument parser

Commit:
cli: Print help of selected command

Printing help in the parse hook doesn't work well,
we need to resolve the selected command first.

jrohel commented:
Looks good. Thanks.
